future-apps/staking/test/staking.js

```diff
@@ -0,0 +1,480 @@
+      const amount = 100
+      const time = 1000
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      const initialBlockNumber = await getBlockNumber()
+      const initialTimestamp = (await getBlock(initialBlockNumber)).timestamp
+      const r = await app.lock(amount / 2, TIME_UNIT_SECONDS, initialTimestamp + time, other, '', '')
+      const lockId = getEvent(r, 'Locked', 'lockId')
+
+      // check lock values
+      const lock = await app.getLock(owner, lockId)
+      assert.equal(lock[0], amount / 2, "locked amount should match")
+      assert.equal(lock[1], TIME_UNIT_SECONDS, "lock time unit should match")
+      assert.equal(lock[2], initialTimestamp + time, "lock time end should match")
+      assert.equal(lock[3], other, "unlocker should match")
+      assert.equal(lock[4], "0x0000000000000000000000000000000000000000000000000000000000000000", "lock metadata should match")
+
+      // can not unlock
+      assert.isFalse(await app.canUnlock(owner, lockId))
+      assert.equal((await app.unlockedBalanceOf(owner)).valueOf(), amount / 2, "Unlocked balance should match")
+      assert.equal((await app.locksCount(owner)).valueOf(), parseInt(lockId, 10) + 1, "last lock id should match")
+
+      await timeTravel(time + 1000)
+      // can unlock
+      assert.isTrue(await app.canUnlock(owner, lockId))
+      // unlockable balance counts as unlocked
+      assert.equal((await app.unlockedBalanceOf(owner)).valueOf(), amount, "Unlocked balance should match")
+
+    })
+
+    it('locks using blocks', async () => {
+      const mockCounter = await getContract('Counter').new()
+      const amount = 100
+      const blocks = 10
+      await token.approve(app.address, amount)
+      await app.stake(amount, '')
+      const initialBlockNumber = await getBlockNumber()
+      const r = await app.lock(amount / 2, TIME_UNIT_BLOCKS, initialBlockNumber + blocks, other, '', '')
+      const lockId = getEvent(r, 'Locked', 'lockId')
+      await timeTravel(2000)
+
+      // check lock values
+      const lock = await app.lastLock(owner)
+      assert.equal(lock[0], amount / 2, "locked amount should match")
+      assert.equal(lock[1], TIME_UNIT_BLOCKS, "lock time unit should match")
+      assert.equal(lock[2], initialBlockNumber + blocks, "lock time end should match")
+      assert.equal(lock[3], other, "unlocker should match")
+      assert.equal(lock[4], "0x0000000000000000000000000000000000000000000000000000000000000000", "lock metadata should match")
+
+      // can not unlock
+      assert.isFalse(await app.canUnlock(owner, lockId))
+      assert.equal((await app.unlockedBalanceOf(owner)).valueOf(), amount / 2, "Unlocked balance should match")
+      assert.equal((await app.locksCount(owner)).valueOf(), parseInt(lockId, 10) + 1, "last lock id should match")
+
+      // we just send dumb transactions to increment block number
+      let blockNumber
+      do {
+        await mockCounter.increment()
```

Comment:
Nice hack 😂🙏 Though this is going to slow down the test unnecessarily. Two possible solutions:

- Make the lock shorter, 2 or 3 blocks should be ok.
- Mock block numbers like we do with timestamps (I thought we did this all over the system, but apparently just in Finance https://github.com/aragon/aragon-apps/blob/master/apps/finance/test/mocks/FinanceMock.sol)

Also rather than introducing a random contract, you can perform ETH value transfers to the same account `web3.eth.sendTransaction({ from: owner, to: owner: value: 1, gasPrice: 1 })` to bump blocks.